api/service.go

```diff
@@ -3,54 +3,58 @@ package api
 
 import (
 	"context"
-	"errors"
 
 	"github.com/heetch/regula"
+	"github.com/heetch/regula/errors"
 	"github.com/heetch/regula/rule"
 )
 
 // API errors.
-var (
-	ErrRulesetNotFound      = errors.New("ruleset not found")
-	ErrRulesetNotModified   = errors.New("not modified")
-	ErrSignatureNotFound    = errors.New("signature not found")
-	ErrInvalidContinueToken = errors.New("invalid continue token")
-	ErrAlreadyExists        = errors.New("already exists")
+const (
+	ErrRulesetNotFound    = errors.Error("ruleset not found")
+	ErrRulesetNotModified = errors.Error("not modified")
+	ErrSignatureNotFound  = errors.Error("signature not found")
+	ErrInvalidCursor      = errors.Error("invalid cursor")
+	ErrAlreadyExists      = errors.Error("already exists")
 )
 
 // RulesetService is a service managing rulesets.
 type RulesetService interface {
 	// Create a ruleset using a signature.
 	Create(ctx context.Context, path string, signature *regula.Signature) error
-	// Put is used to add rules to a ruleset. It creates a new version of the ruleset.
-	Put(ctx context.Context, path string, rules []*rule.Rule) (*regula.Ruleset, error)
-	// Get returns a ruleset alongside its metadata. By default, it returns the latest version.
-	// If the version is not empty, the specified version is returned.
+	// Put is used to add a new version of the rules to a ruleset.
+	Put(ctx context.Context, path string, rules []*rule.Rule) (string, error)
+	// Get returns a ruleset alongside its metadata.
 	Get(ctx context.Context, path, version string) (*regula.Ruleset, error)
-	// List returns the latest version of each ruleset whose path starts by the given prefix.
-	// If the prefix is empty, it returns all the entries following the lexical order.
+	// List returns the list of all rulesets paths.
 	// The listing is paginated and can be customised using the ListOptions type.
-	List(ctx context.Context, prefix string, opt *ListOptions) (*Rulesets, error)
+	List(ctx context.Context, opt ListOptions) (*Rulesets, error)
 	// Watch a prefix for changes and return a list of events.
 	Watch(ctx context.Context, prefix string, revision string) (*RulesetEvents, error)
 	// Eval evaluates a ruleset given a path and a set of parameters. It implements the regula.Evaluator interface.
 	Eval(ctx context.Context, path, version string, params rule.Params) (*regula.EvalResult, error)
 }
 
-// ListOptions contains list options.
-// If the Limit is lower or equal to 0 or greater than 100, it will be set to 50 by default.
+// ListOptions is used to customize the List output.
 type ListOptions struct {
-	Limit         int
-	ContinueToken string
-	PathsOnly     bool // return only the paths of the rulesets
-	AllVersions   bool // return all versions of each rulesets
+	Limit  int    // If the Limit is lower or equal to 0 or greater than 100, it will be set to 50 by default.
```

Comment:
// Limit holds the maximum number of items to return.
// It it's not positive or greater than 100, a limit of 50 will be used.

?
Although it seems slightly odd that I can explicitly specify a limit of 100 and I'll get 100 items, but if I specify 101, I'll only get 50. Why not cap it at 100?

Reply:
Good idea :+1: